File: python/deepstrike/runtime/os_snapshot.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from deepstrike.runtime.kernel_event_log import (
    KernelEventCategory,
    category_for_kind,
    primitive_for_kind,
)

_KERNEL_KINDS = frozenset({
    "compressed",
    "page_out",
    "page_in",
    "large_result_spooled",
    "capability_changed",
    "context_renewed",
    "suspended",
    "resumed",
    "tool_gated",
    "signal_delivery_disposed",
    "budget_exceeded",
    "budget_usage_reported",
    "checkpoint_taken",
    "rollbacked",
    "agent_process_changed",
    "milestone_advanced",
    "milestone_blocked",
    "memory_written",
    "memory_queried",
    "memory_validation_failed",
})


@dataclass
class OsSnapshot:
    last_suspend: dict[str, Any] | None = None
    last_resumed_turn: int | None = None
    process_by_agent: list[dict[str, Any]] = field(default_factory=list)
    budget_exceeded: list[dict[str, Any]] = field(default_factory=list)
    budget_usage_reported: list[dict[str, Any]] = field(default_factory=list)
    signals: list[dict[str, Any]] = field(default_factory=list)
    page_out_count: int = 0
    page_in_count: int = 0
    spool_count: int = 0
    tool_gated_count: int = 0
    memory_written_count: int = 0
    memory_queried_count: int = 0
    memory_validation_failed_count: int = 0
    memory_retrieval_result_count: int = 0


def rebuild_os_snapshot_from_session_events(events: list[dict[str, Any]]) -> OsSnapshot:
    snap = OsSnapshot()
    index: dict[str, int] = {}
    for event in events:
        kind = event.get("kind")
        if kind == "memory_retrieval_result":
            snap.memory_retrieval_result_count += 1
            continue
        if kind not in _KERNEL_KINDS and kind not in ("suspended", "resumed"):
            continue
        if kind == "suspended":
            snap.last_suspend = {
                "turn": event.get("turn"),
                "reason": event.get("reason"),
                "pending_calls": event.get("pending_calls") or [],
            }
        elif kind == "resumed":
            snap.last_resumed_turn = event.get("turn")
        elif kind == "tool_gated":
            snap.tool_gated_count += 1
        elif kind == "agent_process_changed":
            record = {
                "turn": event.get("turn"),
                "agent_id": event.get("agent_id"),
                "parent_session_id": event.get("parent_session_id"),
                "state": event.get("state") or "running",
            }
            agent_id = event.get("agent_id") or ""
            idx = index.get(agent_id)
            if idx is not None:
                snap.process_by_agent[idx] = record
            else:
                index[agent_id] = len(snap.process_by_agent)
                snap.process_by_agent.append(record)
        elif kind == "budget_exceeded":
            snap.budget_exceeded.append({
                "turn": event.get("turn"),
                "operation_id": event.get("operation_id"),
                "reservation_id": event.get("reservation_id"),
                "budget": event.get("budget"),
            })
        elif kind == "budget_usage_reported":
            snap.budget_usage_reported.append({
                "turn": event.get("turn"),
                "operation_id": event.get("operation_id"),
                "reservation_id": event.get("reservation_id"),
                "tokens": event.get("tokens") or 0,
                "subagents": event.get("subagents") or 0,
                "rounds": event.get("rounds") or 0,
            })
        elif kind == "signal_delivery_disposed":
            snap.signals.append({
                "turn": event.get("turn"),
                "operation_id": event.get("operation_id"),
                "delivery_id": event.get("delivery_id"),
                "attempt": event.get("attempt"),
                "signal_id": event.get("signal_id"),
                "disposition": event.get("disposition"),
                "queue_depth": event.get("queue_depth"),
            })
        elif kind == "page_out":
            snap.page_out_count += 1
        elif kind == "page_in":
            snap.page_in_count += 1
        elif kind == "large_result_spooled":
            snap.spool_count += 1
        elif kind == "memory_written":
            snap.memory_written_count += 1
        elif kind == "memory_queried":
            snap.memory_queried_count += 1
        elif kind == "memory_validation_failed":
            snap.memory_validation_failed_count += 1
    return snap

File: node/src/runtime/os-snapshot.ts
import type { SessionEvent } from "./session-log.js"
import { categoryForKind, type KernelEventCategory, type KernelPrimitive, primitiveForKind } from "./kernel-event-log.js"

const KERNEL_KINDS = new Set([
  "compressed",
  "page_out",
  "page_in",
  "large_result_spooled",
  "capability_changed",
  "context_renewed",
  "suspended",
  "resumed",
  "tool_gated",
  "signal_delivery_disposed",
  "budget_exceeded",
  "budget_usage_reported",
  "checkpoint_taken",
  "rollbacked",
  "agent_process_changed",
  "milestone_advanced",
  "milestone_blocked",
  "memory_written",
  "memory_queried",
  "memory_validation_failed",
])

export interface OsSnapshot {
  lastSuspend?: { turn: number; reason: string; pending_calls: string[] }
  lastResumedTurn?: number
  processByAgent: Array<{ turn: number; agent_id: string; parent_session_id: string; state: string }>
  budgetExceeded: Array<{ turn: number; operation_id: string; reservation_id?: string; budget: string }>
  budgetUsageReported: Array<{
    turn: number
    operation_id: string
    reservation_id: string
    tokens: number
    subagents: number
    rounds: number
  }>
  signals: Array<{
    turn: number
    operation_id: string
    delivery_id: string
    attempt: number
    signal_id: string
    disposition: string
    queue_depth: number
  }>
  pageOutCount: number
  pageInCount: number
  spoolCount: number
  toolGatedCount: number
  memoryWrittenCount: number
  memoryQueriedCount: number
  memoryValidationFailedCount: number
  memoryRetrievalResultCount: number
}

export function rebuildOsSnapshotFromSessionEvents(
  events: SessionEvent[],
): OsSnapshot {
  const snap: OsSnapshot = {
    processByAgent: [],
    budgetExceeded: [],
    budgetUsageReported: [],
    signals: [],
    pageOutCount: 0,
    pageInCount: 0,
    spoolCount: 0,
    toolGatedCount: 0,
    memoryWrittenCount: 0,
    memoryQueriedCount: 0,
    memoryValidationFailedCount: 0,
    memoryRetrievalResultCount: 0,
  }
  const index = new Map<string, number>()

  for (const event of events) {
    if (event.kind === "memory_retrieval_result") {
      snap.memoryRetrievalResultCount += 1
      continue
    }
    if (!KERNEL_KINDS.has(event.kind) && event.kind !== "suspended" && event.kind !== "resumed") {
      continue
    }
    switch (event.kind) {
      case "suspended":
        snap.lastSuspend = {
          turn: event.turn,
          reason: event.reason,
          pending_calls: event.pending_calls ?? [],
        }
        break
      case "resumed":
        snap.lastResumedTurn = event.turn
        break
      case "tool_gated":
        snap.toolGatedCount += 1
        break
      case "agent_process_changed": {
        const record = {
          turn: event.turn,
          agent_id: event.agent_id,
          parent_session_id: event.parent_session_id,
          state: event.state ?? "running",
        }
        const idx = index.get(event.agent_id)
        if (idx !== undefined) snap.processByAgent[idx] = record
        else {
          index.set(event.agent_id, snap.processByAgent.length)
          snap.processByAgent.push(record)
        }
        break
      }
      case "budget_exceeded":
        snap.budgetExceeded.push({
          turn: event.turn,
          operation_id: event.operation_id,
          ...(event.reservation_id ? { reservation_id: event.reservation_id } : {}),
          budget: event.budget,
        })
        break
      case "budget_usage_reported":
        snap.budgetUsageReported.push({
          turn: event.turn,
          operation_id: event.operation_id,
          reservation_id: event.reservation_id,
          tokens: event.tokens,
          subagents: event.subagents,
          rounds: event.rounds,
        })
        break
      case "signal_delivery_disposed":
        snap.signals.push({
          turn: event.turn,
          operation_id: event.operation_id,
          delivery_id: event.delivery_id,
          attempt: event.attempt,
          signal_id: event.signal_id,
          disposition: event.disposition,
          queue_depth: event.queue_depth,
        })
        break
      case "page_out":
        snap.pageOutCount += 1
        break
      case "page_in":
        snap.pageInCount += 1
        break
      case "large_result_spooled":
        snap.spoolCount += 1
        break
      case "memory_written":
        snap.memoryWrittenCount += 1
        break
      case "memory_queried":
        snap.memoryQueriedCount += 1
        break
      case "memory_validation_failed":
        snap.memoryValidationFailedCount += 1
        break
      default:
        break
    }
  }
  return snap
}
